feat(pacer): Link pacer sessions to proxies

ERosendo opened the pull request:
This PR introduces a new approach to caching PACER sessions and updates the code of the ProxyPacerSession class to pick one of the proxy connection strings from the settings.

Key Changes:

- Introduces a new environment variable (EGRESS_PROXY_HOSTS) to manage a list of proxies.

- Implements new logic in the ProxyPacerSession  to choose a proxy connection string from the settings, enabling dynamic proxy selection.

- Updates the `log_into_pacer` method to return a tuple containing both the user's cookie and the used proxy address used for login

- Tweaks the `get_or_cache_pacer_cookies` to handle both the new return format and potentially existing cached data in the old format, ensuring backward compatibility.

- Updates tasks from `corpus_importer` and `recap` modules to accommodate the new PACER user cookie format.

- Adapts commands in the `corpus_importer` module that use cookies to handle the new format.

- Introduces new tests for pacer session utils.

This PR closes #4087

mlissner commented on cl/corpus_importer/tasks.py:
```diff
@@ -988,10 +990,21 @@ def get_pacer_case_id_and_title(
+    else:
+        # If the user_pk is unavailable, cookies cannot be retrieved from the
+        # cache.
+        self.request.chain = None
```
Is it better to throw an exception here? In theory we should never arrive at this code, right?

ERosendo replied:
> In theory we should never arrive at this code, right?

You're right! I'll update the code to throw an exception.

mlissner reviewed (changes requested):
Looks like you've got an effective solution here, and that you've thought carefully about how to roll it out without queued tasks failing. Nice. 

I think I would change something though. Having to work with tuples and pass them around is never much fun. They're kind of ugly, it's hard to remember what order the tuple is in ("Was it cookies then proxy or proxy then cookies?"), and so forth. 

Since I think the values we need are getting added to the PACER session object anyway, did you consider just passing that around? Even `log_into_pacer` could just return `s`, I think?

ERosendo commented on cl/recap/tasks.py:
```diff
@@ -1656,8 +1661,13 @@ def fetch_attachment_page(self: Task, fq_pk: int) -> None:
         mark_fq_status(fq, msg, PROCESSING_STATUS.FAILED)
         return
 
+    cookies_data = (
+        cookies
+        if isinstance(cookies, tuple)
+        else (cookies, settings.EGRESS_PROXY_HOST)
+    )
```
I'll add a comment! 👍

ERosendo commented:
@mlissner  I've implemented your suggestions. I've removed the `EGRESS_PROXY_HOST` environment variable from settings and replaced tuples with a dedicated dataclass named `SessionData` for better structure and readability.

Let me know what you think.

mlissner reviewed (approved):
Great! This is a lot easier to reason about. Thanks for making the little helper class.

Can you make an issue with details about deployment and the follow up tasks? E.g.:

 - Need the variables set before deployment
 - Need to remove the old proxies after a few hours or maybe a day
 - Need to remove shim code after a few hours or maybe a day

Anything else?

mlissner commented:
This is approved, but can you work with Ramiro to make sure that we've got the variables and k8s deployments in place before we merge and deploy?

ERosendo commented:
Sure!

ERosendo commented:
@mlissner Ramiro confirmed the environment variable was set up correctly and that CL can successfully reach the newly deployed proxies.

Let's merge this PR. :shipit:

mlissner commented:
Auto-merge is set. Thanks all!

mlissner commented:
@ERosendo Now that this has been live for a couple weeks, is there code in here we want to refactor to simplify things?

ERosendo commented:
@mlissner Yes, We used several ternary if statements to manage legacy cached data. I believe we can safely remove this code. Here are examples:

https://github.com/freelawproject/courtlistener/blob/57162841fd5aeda54bab59ee0b8439652fcda6e6/cl/recap/tasks.py#L1658-L1660

https://github.com/freelawproject/courtlistener/blob/57162841fd5aeda54bab59ee0b8439652fcda6e6/cl/recap/tasks.py#L2416-L2421

https://github.com/freelawproject/courtlistener/blob/57162841fd5aeda54bab59ee0b8439652fcda6e6/cl/lib/pacer_session.py#L165-L169
